tesla_api/energy.py

```diff
@@ -50,10 +59,39 @@ async def get_battery_count(self):
         info = await self.get_energy_site_info()
         return int(info["battery_count"])
 
-
     async def get_energy_site_live_status(self):
         return await self._api_client.get('energy_sites/{}/live_status'.format(self._energy_site_id))
 
+    async def get_energy_site_calendar_history_data(self, kind: str = 'energy', period: str = 'day',
+                                                    end_date: [str, datetime] = None,
+                                                    time_zone: str = None) -> Optional[dict]:
+        """
+        Returns historical energy data
+        :param kind: [power, energy, self_consumption]
+        :param period: Amount of time to include in report. One of day, week, month, year, and lifetime.
+                    When kind is 'power', this parameter is ignored, and the period is always 'day'
+        :param end_date: ISO 8601 datetime, e.g. 2019-12-23T17:39:18.546Z. either in str or as datetime
+                    The response report interval ends on this datetime and starts at the beginning of the given period
+                    at 1:00 AM.  Note time cannot be 00:00. Defaults to the current time.
+        :param time_zone: IANA/Olsen time zone identifier, e.g. America/New_York. If set, and if the end_date is a
+                    'naive' datetime object (without timezone info) this field will be used to set the timezone
+                    for the end_date.  Otherwise it is ignored
+        """
+        params = {'kind': kind, 'period': period}
+        if end_date is not None:
+            if type(end_date) is datetime:
+                if end_date.hour == 0:
+                    # if the datetime object's time is 00:00 then the API returns nothing. We adjust by adding an hour
+                    # so you can use datetime(year=2020, month=5, day=2) and it gets the data for May 2, 2020 as expected
+                    end_date = end_date + timedelta(hours=1)
+                if end_date.tzinfo is None and time_zone is not None:
+                    tz = pytz.timezone(time_zone)
```

Comment:
Will `datetime.timezone` not work here? pytz is not part of the standard library, so probably don't want to add it as a dependency for this one line.

Reply:
There are a few issues with datetime.timezone:

1. It does not contain the IANA database for timezone conversions, yet that is the standard used by the Tesla API.  From the python documentation on the datetime.timezone class:
> dateutil.tz library brings the IANA timezone database (also known as the Olson database) to Python, and its usage is recommended
From dateutil is I believe heavier than pytz and from stackoverflow   I understand pytz to be the preferred way to add IANA timezones 
2.  it takes a fixed time offset, whereas pytz correctly applies daylight savings and changes to the timezone over time (e.g., a state deciding to no longer use DST).  Per [this](https://stackoverflow.com/questions/18541051/datetime-and-timezone-conversion-with-pytz-mind-blowing-behaviour) and various stackoverflow articles it is recommended to use 

The alternative is to expect the datetime object to have the timezone set already, but that simply moves the problem back to the user, whereas the IANA timezone is spit out by the Tesla API already when querying the site, so it's easy to add it.  The typical use of the `get_energy_site_calendar_history_data` method becomes:
`response = tesla_client.power_for_day(site_id=site_id, day=datetime(year=2020, month=5, day=1), time_zone='America/Los_Angeles')`

Reply:
Sorry I added my call to a wrapper around the API.  The call for power for a day becomes:
`response = await site.get_energy_site_calendar_history_data(kind='power', end_date=day, time_zone=time_zone)`
with day simply a datetime like `datetime(year=2020, month=5, day=1)` and the time_zone an IANA string (obtained from the API) such as `time_zone='America/Los_Angeles'`

Reply:
I'm failing to find any documentation on these APIs, and can't test it myself. So, just a couple of questions to make sure I understand this correctly.

The timezone actually comes from another endpoint? What code would you write to get that timezone info?

With period set to day, this retrieves the data for a day preceding the given date, but the starting time will always be 01:00 of the preceding day? So, does this go up to the start of the given day, or up to the exact time passed? e.g. If I set the date as 6th 23:59, does that give data for 47 hours (5th 01:00 till 6th 23:59), or does it completely ignore the time and just do till the beginning of the day (5th 01:00 till 6th 01:00)?

Reply:
I don't think there is formal API documentation (because it is not an officially sanctioned API by Tesla).  The best documentation I have found is [here](https://www.teslaapi.info/energy-sites/state-and-settings#historical-calendar-data), and I have experimented a bit myself.  The timezone data is obtained like this:
`response = await site.get_energy_site_calendar_history_data(kind='power')
            if 'installation_time_zone' in response:
                self.time_zones[site_id] = response['installation_time_zone']`
i.e. calling the api without supplying a date.  Then, you can get the power for one day by calling:
`response = await site.get_energy_site_calendar_history_data(kind='power', end_date=day,
                                                                        time_zone=time_zone)
`
and the reply has a key 'time_series' which returns the power every 15 mins, starting at 0:00.  I don't know why you can't pass a datetime with the hour set to 00:00, but you can't: it doesn't return an error but an empty string - very strange.

Reply:
Ok, hold on for a minute as I need to check something else with this weird API

Reply:
OK, that documentation says that timezone seems to have no effect, so seems there's no point in including it.

So, in my example it would give 47 hours of data in 15 min intervals, right? Weird.

Reply:
Ok, in the latest push I've deleted the timezone argument altogether.  This means you now have to ensure that the datetime object you pass as end date is timezone aware, otherwise you'll get an error.  I've also changed the increment such that when you pass as end date a datetime without a time (e.g. datetime(year=220, month=5, day=1) you get all data for May 1st as you would expect.  
The raw API call will return data up until the end_date, so if you enter May 1st at 3pm you'll get all data on May 1st from 00:00 to 15:00 in 5 minute intervals.  Normally you would call with a date only, and then get the data for that date (only).  If you call without an end date you get the data for today, from 00:00 to now